packages/tree/src/node.rs

```diff
@@ -46,13 +51,59 @@ impl<'r, T> NodeRef<'r, T> {
 
             Some(NodeRef {
                 tree: self.tree,
-                this: this.parent(),
+                this: this.parent_index(),
             })
         }
     }
 
+    #[inline]
+    pub fn parents_iter(&self) -> ParentsIter<'r, T> {
+        ParentsIter { node: *self }
+    }
+
+    #[inline]
+    /// This exists as a functional approach to converting node reference into
+    /// a subtree.
+    pub fn to_subtree(&self) -> Subtree<'r, T> {
+        Subtree::from_node(*self)
+    }
+
+    #[inline]
+    pub(crate) fn tree(&self) -> &'r Tree<T> {
+        self.tree
+    }
+
+    #[inline]
+    pub(crate) fn this_index(&self) -> u16 {
+        self.this
+    }
+
     #[inline]
     pub(crate) const fn with_index(tree: &'r Tree<T>, this: NodeIndex) -> Self {
         Self { tree, this }
     }
 }
+
+impl<'r, T> Clone for NodeRef<'r, T> {
+    fn clone(&self) -> Self {
+        Self { ..*self }
+    }
+}
+
+impl<'r, T> Copy for NodeRef<'r, T> {}
+
+pub struct ParentsIter<'r, T> {
+    node: NodeRef<'r, T>,
+}
+
+impl<'r, T> Iterator for ParentsIter<'r, T> {
+    type Item = NodeRef<'r, T>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let parent = self.node.parent()?;
```

Comment:
this is unstable yet https://doc.rust-lang.org/src/core/option.rs.html#2285

Reply:
how does it compile?

Reply:
As discussed in DMs, this is just a sugar. It's usable with both Result and Option types.